### ISSUE TITLE
Temporal regularization experiment with electricity data

### ISSUE THREAD

**Issue by deploy-soon** [deploy-soon/TemporalMF#2]
Some experiment results which contain validation loss among several loss functions.
Electricity data is from [Multivariate Time series Data sets](https://github.com/laiguokun/multivariate-time-series-data).

**Electricity**

The raw dataset is in https://archive.ics.uci.edu/ml/datasets/ElectricityLoadDiagrams20112014. It is the electricity consumption in kWh was recorded every 15 minutes from 2011 to 2014. Because the some dimensions are equal to 0. So we eliminate the records in 2011. Final we get data contains electircity consumption of 321 clients from 2012 to 2014. And we converted the data to reflect hourly consumption.

shape: (26304, 321)

**Comment by deploy-soon** [comment]
Base Matrix Factorization
- code: [`10d386`](https://github.com/deploy-soon/TemporalMF/commit/10d386605e47fb931e67027e7b48c5058d2ac242)
- loss function
![image](https://user-images.githubusercontent.com/44163510/75543102-71298f80-5a64-11ea-8e07-8c311f1d8b3f.png)
- Adam optimizer(learning rate=0.0005)
- train ratio: 0.8
- metric:
![image](https://user-images.githubusercontent.com/44163510/75543454-29573800-5a65-11ea-81f1-fe1e96da7eaf.png)
- result

|**Factors** | 20     | 30     | 40     | 50     |
|---------|--------|--------|--------|--------|
| **NRMSE**   | 1.824 | 1.628 | 1.539 | 1.365 |

**Comment by deploy-soon** [comment]
L2 Reg Matrix Factorization
- code: [`10d386`](https://github.com/deploy-soon/TemporalMF/commit/10d386605e47fb931e67027e7b48c5058d2ac242)
- loss function
![image](https://user-images.githubusercontent.com/44163510/75551705-1ef26980-5a78-11ea-9d07-3e4c9404aab9.png)
- Adam optimizer(learning rate=0.005)
- train ratio: 0.8
- metric:
![image](https://user-images.githubusercontent.com/44163510/75543454-29573800-5a65-11ea-81f1-fe1e96da7eaf.png)
- result

| Factors      | 20     | 30     | 40     | 50 |
|--------------|--------|--------|--------|----|
| lambda=5.0   |  1.790 | 1.551 | 1.426  | 1.266  |
| lambda=0.5   | 1.820 | 1.622 | 1.500 |  1.390  |
| lambda=0.05  | 1.842 | 1.643 | 1.489 |  1.423  |
| lambda=0.005 | 1.870 | 1.628 | 1.498 |  1.383  |

**Comment by deploy-soon** [comment]
Temporal L2 Reg Matrix Factorization
- code: [`10d386`](https://github.com/deploy-soon/TemporalMF/commit/10d386605e47fb931e67027e7b48c5058d2ac242)
- loss function
![image](https://user-images.githubusercontent.com/44163510/75846026-4d789780-5e1e-11ea-8cf7-fb51e509ffa6.png)
- Adam optimizer(learning rate=0.0005)
- train ratio: 0.8
- metric:
![image](https://user-images.githubusercontent.com/44163510/75543454-29573800-5a65-11ea-81f1-fe1e96da7eaf.png)
**result**
- temporal lambda = lambda / 10.0

| Factors      | 20     | 30     | 40     | 50 |
|--------------|--------|--------|--------|----|
| lambda=5.0   | 1.497  | 1.266  | 1.119  |  1.010  |
| lambda=0.5   | 1.557 | 1.346 | 1.178 |  1.098  |
| lambda=0.05  | 1.610 | 1.397 | 1.263 | 1.114   |
| lambda=0.005 | 1.563 | 1.408 | 1.255 |  1.164  |

  - temporal lambda = lambda / 1.0

| Factors      | 20     | 30     | 40     | 50 |
|--------------|--------|--------|--------|----|
| lambda=5.0   | 1.511  | 1.292  | 1.099  |  1.043  |
| lambda=0.5   | 1.566 | 1.332 | 1.201 |  1.088  |
| lambda=0.05  | 1.584 |1.365  | 1.212 |  1.093  |
| lambda=0.005 | 1.615 | 1.394 | 1.234 |  1.203  |


  - temporal lambda = lambda / 0.1

| Factors      | 20     | 30     | 40     | 50 |
|--------------|--------|--------|--------|----|
| lambda=5.0   | 1.512  | 1.312  | 1.109  |  0.9400  |
| lambda=0.5   | 1.562 | 1.366 | 1.143 |  1.062  |
| lambda=0.05  | 1.593 | 1.387 | 1.272 | 1.0986   |
| lambda=0.005 | 1.564 | 1.395 | 1.242 |  1.144  |


  - temporal lambda = lambda / 0.01

| Factors      | 20     | 30     | 40     | 50 |
|--------------|--------|--------|--------|----|
| lambda=5.0   |  1.500 |  1.197 |  0.9403 |  0.8244  |
| lambda=0.5   | 1.520 | 1.329 | 1.125 | 0.9742   |
| lambda=0.05  | 1.602 | 1.424 | 1.208 |  1.081  |
| lambda=0.005 | 1.563 | 1.457 | 1.220 |  1.116  |

**Comment by deploy-soon** [comment]
The result between the model which is trained with difference of temporal embedding and the model which is predicted by just temporal embedding.

|table|factors=40 lags=5 is_pred_sub=False|factors=40 lags=5 is_pred_sub=True|factors=40 lags=50 is_pred_sub=False|factors=40 lags=50 is_pred_sub=True|factors=40 lags=100 is_pred_sub=False|factors=40 lags=100 is_pred_sub=True|factors=60 lags=5 is_pred_sub=False|factors=60 lags=5 is_pred_sub=True|factors=60 lags=50 is_pred_sub=False|factors=60 lags=50 is_pred_sub=True|factors=60 lags=100 is_pred_sub=False|factors=60 lags=100 is_pred_sub=True|
|---|---|---|---|---|---|---|---|---|---|---|---|---|
|name=MatrixMF|0.5332|0.4989|0.4911|0.4855|0.4958|0.4871|0.4956|0.4828|0.489|0.4734|0.4837|0.4824|
|name=LSTMMF|0.4338|0.4362|0.4417|0.4282|0.4452|0.4286|0.4307|0.4357|0.4424|0.4178|0.4323|0.4211|

All cases told that the model trained with the temporal embedding difference is better than naive one.

**Comment by deploy-soon** [comment]
Test loss for several recurrent layers and dimensions just LSTM.

- dropout rate: 0.3
- lambda_x = 5.0
- lags: 100

|LSTM|factors=20|factors=40|factors=80|
|---|---|---|---|
|n_layers=1 hid_dim=128|0.4704 611880|0.4471 1157200|0.4535 2247840|
|n_layers=1 hid_dim=256|0.4693 822312|0.4362 1380432|0.4305 2496672|
|n_layers=2 hid_dim=128|0.4734 743976|0.4589 1289296|0.4171 2379936|
|n_layers=2 hid_dim=256|0.4709 1348648|0.4311 1906768|0.4594 3023008|

**Comment by deploy-soon** [comment]
#### Forecasting with Full Observations
- validation: 24 * 7 days
- Matrix Embeddings
NRMSE of validation loss.

| NRMSE | factors=20 | factors=40 | factors=80 |
|:-----:|------------|------------|------------|
|       | 1.3877     | 1.1551     | 1.2431     |

**Comment by deploy-soon** [comment]
**Missing Imputations**

- train ratio: 0.7, validation test ratio: 0.15 each
- lags: 100
- lstm with 1 layer.
- learning_rate: 0.005
- LSTNet without attention
- test loss using normalization.

|      NRMSE     | factors=40 | factors=80 |
|:--------------:|------------|------------|
| Base Attention | 0.4465     | 0.4637     |
|     LSTNet     | 0.4617     | 0.4842     |
|       TPA      | 0.4471     | 0.4418     |
